crates/rpc/src/v04/method/trace_transaction.rs

```diff
@@ -0,0 +1,173 @@
+            ExecutionStateError::Internal(e) => Self::Internal(e),
+        }
+    }
+}
+
+impl From<CallError> for TraceTransactionError {
+    fn from(value: CallError) -> Self {
+        match value {
+            CallError::ContractNotFound | CallError::InvalidMessageSelector => {
+                Self::NoTraceAvailable(TraceError::Rejected)
+            }
+            CallError::Reverted(_) => Self::NoTraceAvailable(TraceError::Received),
+            CallError::Internal(e) => Self::Internal(e),
+        }
+    }
+}
+
+impl From<JoinError> for TraceTransactionError {
+    fn from(e: JoinError) -> Self {
+        Self::Internal(anyhow::anyhow!("Join error: {e}"))
+    }
+}
+
+impl From<anyhow::Error> for TraceTransactionError {
+    fn from(e: anyhow::Error) -> Self {
+        Self::Internal(e)
+    }
+}
+
+impl From<TraceTransactionError> for RpcError {
+    fn from(value: TraceTransactionError) -> Self {
+        match value {
+            TraceTransactionError::InvalidTxnHash => RpcError::InvalidTxnHash,
+            TraceTransactionError::NoTraceAvailable(status) => RpcError::NoTraceAvailable(status),
+            TraceTransactionError::Internal(e) => RpcError::Internal(e),
+        }
+    }
+}
+
+pub async fn trace_transaction(
+    context: RpcContext,
+    input: TraceTransactionInput,
+) -> Result<TraceTransactionOutput, TraceTransactionError> {
+    let (transactions, block_hash, gas_price): (Vec<Transaction>, BlockHash, Option<U256>) = {
+        let span = tracing::Span::current();
+
+        let storage = context.storage.clone();
+        tokio::task::spawn_blocking(move || {
+            let _g = span.enter();
+
+            let mut db = storage.connection()?;
+            let tx = db.transaction()?;
+
+            let block_hash = tx
+                .transaction_block_hash(input.transaction_hash)?
+                .ok_or(TraceTransactionError::InvalidTxnHash)?;
+
+            let gas_price: Option<U256> = tx
+                .block_header(pathfinder_storage::BlockId::Hash(block_hash))?
+                .map(|header| U256::from(header.gas_price.0));
+
+            let (transactions, _): (Vec<_>, Vec<_>) = tx
+                .transaction_data_for_block(pathfinder_storage::BlockId::Hash(block_hash))?
+                .ok_or(TraceTransactionError::InvalidTxnHash)?
+                .into_iter()
+                .unzip();
+
+            let transactions = transactions
+                .into_iter()
+                .map(|transaction| compose_executor_transaction(transaction, &tx))
+                .collect::<anyhow::Result<Vec<_>, _>>()?;
+
+            Ok::<_, TraceTransactionError>((transactions, block_hash, gas_price))
+        })
+        .await
+        .context("trace_transaction: fetch & map the transaction")??
+    };
+
+    let block_id = BlockId::Hash(block_hash);
+    let execution_state = crate::executor::execution_state(context, block_id, gas_price).await?;
```

Comment:
This is not _quite_ right I think. To execute transactions in `block_id` you'd actually need _state_ from the previous block (since state at block X means state _after_ all transactions in the block has been executed).